### PR TITLE
sys: add common imath module mv isin() form test/driver_dac_dds

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

An integer-only `sin()` function is generally useful, move it to a common place `imath` where more integer-only math functions can be added when needed.

### Testing procedure

This only moves code around.

### CODE REVIEW

File: sys/imath/imath.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include "imath.h"
+
+/**
+ * @brief A sine approximation via a fourth-order cosine approx.
+ *        source: https://www.coranac.com/2009/07/sines/
+ *
+ * @param x     angle (with 2^15 units/circle)
+ * @return      sine value (Q12)
+ */
+int32_t isin(int32_t x)
```
Comment (kfessel) [inline sys/imath/imath.c:26]: wouldn't it fit to be put static in the header?

Comment (benpicco) [reply sys/imath/imath.c:26]: You mean the whole function as a `static inline`?
Why, it's more than a small helper function, I wouldn't want to have the code duplicated everywhere.

Comment (kfessel) [reply sys/imath/imath.c:26]: I somehow think its more likely to have on compile unit and one function that uses isine -> saves memory to have it inlined.

Comment (benpicco) [reply sys/imath/imath.c]: Well if you insist ¯\\_(ツ)_/¯

**Review by kfessel — COMMENTED** [review]
I stared at this for too long:
bike-shed:
- some pre- or postfix for the function names that indicate that it is not only special for its type but also less accurate i think `fast_isin()` would be nice (like fast_sqrt (just realised we don't have that quake algorithm :) ).
- ISIN_PERIOD is wrong 

 https://godbolt.org/z/edqv4T6YM

this get us both fast_isin and fast_icos  (icos being smaller) 
`iPI = 0x8000>>1`

```
static inline  __attribute__((always_inline))
int32_t ihelp(int32_t x){
    int32_t y;
    static const int32_t qN = 13,
                         qA = 12,
                          B = 19900,
                          C = 3516;

    x = x << (31 - qN);         /* Mask with PI                 */
    x = x >> (31 - qN);         /* Note: SIGNED shift! (to qN)  */
    x = x * x >> (2 * qN - 14); /* x=x^2 To Q14                 */

    y = B - (x * C >> 14);      /* B - x^2*C                    */
    y = (1 << qA)               /* A - x^2*(B-x^2*C)            */
      - (x * y >> 16);

    return y;
}

//static inline
int32_t isin(int32_t x){
    static const int32_t qN = 13;
    int32_t c = (x << (30 - qN));
    int32_t y = ihelp(x - (1 << qN));
    return c >= 0 ? y :-y ;
}

//static inline
int32_t icos(int32_t x){
    static const int32_t qN = 13;
    // vvv this is most probalbly not the optimal way to write that
    //     but the compiler gets it
    int32_t c = ((x + iPI/2) << (30 - qN));  
    int32_t y = ihelp(x);
    return c >= 0 ? y :-y ;
}
```

**Comment by benpicco** [comment]
> `iPI = 0x8000>>1`

so `iPI = ISIN_PERIOD >> 1`?

**Comment by benpicco** [comment]
> it is not only special for its type but also less accurate

but is it really less accurate if we only have fixed point? The accuracy is pretty much given by the number format.

Plot of the values looks good.

isin

I did drop the ambiguous Q notation entirely.

**Comment by maribu** [comment]
@kfessel asked me to join the bike shedding to untie the knot. My two cents are:

- there are already `double sin(double x)`, `float sinf(float x)`, `long double sinl(long double x)` in the C standard, so an `int32_t sini(int32_t x)` (or maybe even an `int32_t sini32(int32_t x)`) would be more consistent
- if there is an accuracy vs speed trade of an `int32_t fastsini32(int32_t x)` would make this obvious. But so would also be a note in Doxygen. Maybe a compromise would be an `static inline` wrapper without the `fast` that calls the fast one. And if one explicitly wants a fast or a highly accurate implementation, one could use the long name.

But to be honest, I'm not much a fan of bike shedding. I'm usually happy with whatever name the person investing the time and effort in PR picked for naming. IMO time is better spent on the implementation than on the naming of functions or data types.

**Comment by benpicco** [comment]
Ok, I think it looks more ugly now, but there is now consistency with `math.h`

**Review by kfessel — APPROVED** [review]

**Comment by kfessel** [comment]
bors merge

**Comment by maribu** [comment]
bors cancel

**Comment by maribu** [comment]
bors cancel

**Comment by maribu** [comment]
bors merge

**Pull request merged**